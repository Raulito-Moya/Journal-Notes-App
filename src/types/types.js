
export const types = {
  
  //estas son el tipo de accion para el login
  login: '[Auth] login',
  logout: '[Auth] logout',  

  // estas son el tipo de accion para el error del form
  uiSetError: '[UI] Set Error',
  uiRemoveError: '[UI] Remove Error',

  uiStartLoading: '[UI] Start loading',
  uiFinishLoading: '[UI] Finish loading',


  notesAddNew: '[Notes] New note',
  notesActive: '[Notes] Set active note',
  notesLoad: '[Notes] Load Note',
  notesUpdated: '[Notes] Updated note saved',
  notesFileUrl: '[Notes] Updated image url',
  notesDelete: '[Notes] Delete note',
  notesLogoutCleaning: '[Notes] Logout Cleaning'
}