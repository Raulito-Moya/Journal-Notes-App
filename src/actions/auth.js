import Swal from 'sweetalert2'
import withReactContent from 'sweetalert2-react-content'

import {firebase, googleAuthProvider} from '../firebase/firebase-config';
import { types } from "../types/types";
import { noteLogout } from './notes';
import { finishLoading, startLoading } from './ui';
 
const MySwal = withReactContent(Swal) //para los efectos de aniamacion en el form

export const startLoginEmailPassword = (email,password) => { //aqui voy a simular una accion asincrona
     return (dispatch) => {
        
       dispatch(startLoading())
     return  firebase.auth().signInWithEmailAndPassword(email,password) //esta instruccion es para logear al usuario
              .then( ({user}) => {
                 // console.log(user)
               dispatch( login(user.uid, user.displayName))  
               dispatch(finishLoading())
            })
            .catch(e => {
               console.log(e); 
               dispatch(finishLoading()) //por si hay un error
               MySwal.fire( 'Error', e.message , 'error')
              })

     }

};

export const startRegisterRegisterWithEmailPasswordName = ( email, password, name) => {

   return( dispatch ) => {

      firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async({user}) =>{

            await user.updateProfile({ displayName: name }) //update user profile data, se espera que se actualize el objeto para despues hacr el dispatch

            //  console.log(user);
                  dispatch( 
                       login(user.uid, user.displayName) 
                      ) 
             })
             .catch(e =>{
               console.log(e);
               MySwal.fire( 'Error', e.message , 'error')
             })
   }



};

export const startGoogleLogin = () => {  //aqui hago un sign in con google
   return (dispatch) => {
  
     firebase.auth().signInWithPopup(googleAuthProvider)
           .then( ({user}) =>{
             dispatch( login(user.uid, user.displayName) ) 
           })
   }


};

export const login = (uid, displayName) => ({
       type: types.login,
       payload: {
         uid,
         displayName
       }
});


export const startLogout = () => {
     return async( dispatch ) => {

      await firebase.auth().signOut();
      dispatch(logout())
      dispatch(noteLogout()) //
     }

}


export const logout = () => ({
   type: types.logout

})