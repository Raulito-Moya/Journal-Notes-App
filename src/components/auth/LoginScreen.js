import React from 'react';
import {Link} from 'react-router-dom';
import {useDispatch, useSelector} from 'react-redux';
import { login, startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'; //aqui traigo la accion
import {setError, removeError} from '../../actions/ui'
import validator from 'validator';
import { useForm } from '../../hooks/useForm';


export const LoginScreen = () => {

     const dispacth = useDispatch()   //el useDispatch es para hacer el dispatch de acciones
     const {msgError, loading} = useSelector(state=> state.ui)
    
    // console.log(state);


     const [formValues, handleInputChange, reset] = useForm({
       email: 'nando8moya@gmail.com',
       password: '123456'
     });

    const {email,password} = formValues;



    const handleLogin = (e) =>{
      e.preventDefault()

      
         dispacth(startLoginEmailPassword(email,password)) //aqui va a disparar la accion asincrona
         //console.log(email, password);
      
    };


    const handleGoogleLogin = () => {
    
    if(isFormValid()){ 
       dispacth(startGoogleLogin()) 
      }  
      console.log('hey!!');
    };




    const isFormValid = () => {

        if( !validator.isEmail(email)){
         dispacth(setError('email is not valid'))
          console.log('email is not valid')
          return false;
   
      } else if( password.length < 6  ){
         dispacth(setError('wrong password'))
        console.log('Password should be at least 6 charecters and match each other');
        return false
      }
       
      dispacth(removeError())
      return true 
     
    }
   

 return(
     <>
         <h3 className="auth__title">Login</h3>

         <form 
         onSubmit={handleLogin}
        className="animate__animated animate__fadeIn animate__faster"  >

           { 
            msgError && ( 
             <div className="auth__alert-error">
               {msgError}
           </div>)
           
           }  

             <input 
             type="text" 
             placeholder="Email"
             name="email" 
             value={email}
             className="auth__input"
             autoComplete="off"
             onChange={handleInputChange}/>

            <input 
             type="password" 
             placeholder="password"
             name="password" 
             value={password}
              className="auth__input"
              onChange={handleInputChange}/>

         
             <button 
               type="submit"
               className="btn btn-primary btn-block"
               disabled={loading}
               >
                Login
            </button>

            <hr/>
            
           <div className="auth_social-networks">
               <p>Login with social networks</p>

               <div 
                  className="google-btn"
                  onClick={handleGoogleLogin}
                 >
               <div className="google-icon-wrapper">
                  <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
               </div>
                 <p className="btn-text">
                   <b>Sign in with google</b>
                </p>
               </div>
           </div>
   
           <Link 
           to='/auth/register'
           className="link"
           >
             Create New Acount  
           </Link>
            
         </form>
     </>
 )



}