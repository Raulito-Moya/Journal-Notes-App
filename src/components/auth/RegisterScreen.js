import React from 'react';
import {useDispatch, useSelector} from 'react-redux'
import {Link} from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import validator  from 'validator'; //Aqui implemento validator paa validar los campos
import {setError,removeError} from '../../actions/ui'
import { startRegisterRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen =()=>{

  const dispatch = useDispatch();
  const { msgError } = useSelector( state => state.ui) //el useSelector es para seleccionar el estado actual y usarlo en la vista

  console.log(msgError);

  const [formValues, handleInputChange, reset] = useForm({
     name:'Nando',
     email:'nando8moya@gmail.com',
     password:'123456',
     password2:'123456'
   })

 const {name,email,password,password2} =formValues;

  
 const handleRegister = (e) => {
    e.preventDefault()
   
    if (isFormValid()) {
      dispatch(startRegisterRegisterWithEmailPasswordName(email, password, name));
    }

 }

 const isFormValid = () => {

   if(name.trim().length === 0 ) {
      dispatch(setError('name is required'))
      console.log('name is required')
      return false;

   } else if( !validator.isEmail(email)){
      dispatch(setError('email is not valid'))
       console.log('email is not valid')
       return false;

   } else if( password !== password2 || password.length < 5){
      dispatch(setError('Password should be at least 6 charecters and match each other'))
     console.log('Password should be at least 6 charecters and match each other');
     return false
   }
    
   dispatch(removeError())
   return true 
  
 }







   return(
    <div>
         
      <h3 className="auth__title">Register</h3>

       <form 
        onSubmit={handleRegister}
        className="animate__animated animate__fadeIn animate__faster">

       { 
         msgError && 
        ( 
          <div className="auth__alert-error">
             {msgError}
          </div>
         )
        }

           <input 
           type="text" 
           placeholder="name"
           name="name" 
           value={name}
           onChange={handleInputChange}
           className="auth__input"
           autoComplete="off"/>
     
          <input 
           type="text" 
           placeholder="email"
           name="email" 
           value={email}
           onChange={handleInputChange}
           className="auth__input"
           autoComplete="off"/>
  

      
          <input 
              type="password" 
              placeholder="password"
              name="password" 
              value={password}
              onChange={handleInputChange}
               className="auth__input"/>
          

             <input 
              type="password" 
              placeholder="Confirm Password" 
              name="password2" 
              value={password2}
              onChange={handleInputChange}
              className="auth__input"/>
          
              <button 
                type="submit"
                className="btn btn-primary btn-block mb-5"
                disabled={false}
                >
              Register
             </button>
          
          
          
            <Link 
            to='/auth/login'
            className="link"
            >
              Alredy Registered
            </Link>
             
          </form>
                 </div>
             )
  
  
}