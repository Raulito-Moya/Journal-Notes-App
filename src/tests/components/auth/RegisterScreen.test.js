import React from 'react'
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import {RegisterScreen} from '../../../components/auth/RegisterScreen'
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { types } from '../../../types/types';



const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initialState = {
    auth: {},
    ui: {
       loading: false,
       msgError: null
    }
}

let store = mockStore(initialState)
//store.dispatch =jest.fn()


describe('Pruebas en el <RegisterScreen/>', () => {
      
  /*  beforeEach(()=>{
        store= mockStore(initialState)
    })*/


    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
           <RegisterScreen/>
        </MemoryRouter>
      </Provider>
    )



  test('debe mostrarse correctamente',()=>{

     
     expect(wrapper).toMatchSnapshot()

  })


  test('debe hacer el dispatch de la accion respectiva', ()=>{

      const emailField = wrapper.find('input[name="email"]')
   
     // console.log(emailField.exists());
   
      emailField.simulate('change',{
          target: {
              value: '',
              name: 'email'
          }
      });
   
      wrapper.find('form').simulate('submit',{
          preventDefault(){}
      });
   
      const actions = store.getActions();
     // console.log(actions);
   
        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'email is not valid'
        }); 
       
  })


  test('debe de mostrar la caja de alerta con el error',() => {
     
    const initialState = {
      auth: {},
      ui: {
         loading: false,
         msgError: 'Email no es correcto'
      }
    }
    
    let store = mockStore(initialState)

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
           <RegisterScreen/>
        </MemoryRouter>
      </Provider>
     )

       expect(wrapper.find('.auth__alert-error').exists()).toBe(true)
       expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initialState.ui.msgError)
  })


});