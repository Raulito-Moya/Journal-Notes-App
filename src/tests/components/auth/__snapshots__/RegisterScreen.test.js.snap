// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <RegisterScreen/> debe mostrarse correctamente 1`] = `
<div>
  <h3
    className="auth__title"
  >
    Register
  </h3>
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="name"
      type="text"
      value="Nando"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="email"
      type="text"
      value="nando8moya@gmail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm Password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      disabled={false}
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Alredy Registered
    </a>
  </form>
</div>
`;
